nav2_smoother/src/nav2_smoother.cpp

```diff
@@ -0,0 +1,298 @@
+    it->second->deactivate();
+  }
+  plan_publisher_->on_deactivate();
+
+  // destroy bond connection
+  destroyBond();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+SmootherServer::on_cleanup(const rclcpp_lifecycle::State &) {
+  RCLCPP_INFO(get_logger(), "Cleaning up");
+
+  // Cleanup the helper classes
+  SmootherMap::iterator it;
+  for (it = smoothers_.begin(); it != smoothers_.end(); ++it) {
+    it->second->cleanup();
+  }
+  smoothers_.clear();
+
+  // Release any allocated resources
+  action_server_.reset();
+  plan_publisher_.reset();
+  tf_.reset();
+  transform_listener_.reset();
+  footprint_sub_.reset();
+  costmap_sub_.reset();
+  collision_checker_.reset();
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+SmootherServer::on_shutdown(const rclcpp_lifecycle::State &) {
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+bool SmootherServer::findSmootherId(const std::string &c_name,
+                                    std::string &current_smoother) {
+  if (smoothers_.find(c_name) == smoothers_.end()) {
+    if (smoothers_.size() == 1 && c_name.empty()) {
+      RCLCPP_WARN_ONCE(get_logger(),
+                       "No smoother was specified in action call."
+                       " Server will use only plugin loaded %s. "
+                       "This warning will appear once.",
+                       smoother_ids_concat_.c_str());
+      current_smoother = smoothers_.begin()->first;
+    } else {
+      RCLCPP_ERROR(get_logger(),
+                   "SmoothPath called with smoother name %s, "
+                   "which does not exist. Available smoothers are: %s.",
+                   c_name.c_str(), smoother_ids_concat_.c_str());
+      return false;
+    }
+  } else {
+    RCLCPP_DEBUG(get_logger(), "Selected smoother: %s.", c_name.c_str());
+    current_smoother = c_name;
+  }
+
+  return true;
+}
+
+void SmootherServer::smoothPlan() {
+  RCLCPP_INFO(get_logger(), "Received a goal, smoothing.");
+
+  auto result = std::make_shared<Action::Result>();
+  try {
+    std::string c_name = action_server_->get_current_goal()->smoother_id;
+    std::string current_smoother;
+    if (findSmootherId(c_name, current_smoother)) {
+      current_smoother_ = current_smoother;
+    } else {
+      action_server_->terminate_current();
+      return;
+    }
+
+    // Perform smoothing
+    auto goal = action_server_->get_current_goal();
+    result->path = goal->path;
+    auto start_time = steady_clock_.now();
+    result->was_completed = smoothers_[current_smoother_]->smooth(
+        result->path, goal->max_smoothing_duration);
+    result->smoothing_duration = steady_clock_.now() - start_time;
+
+    plan_publisher_->publish(result->path);
+
+    RCLCPP_INFO(get_logger(), "Collision check");
+
+    // Check for collisions
+    geometry_msgs::msg::Pose2D pose2d;
```

Comment:
Add a parameter that defaults to false to check if the output path contains collisions. There may be situations where that's not desirable (like when we are using a smoother that isn't cost aware, or without costmaps in the frame, or when collisions are acceptable in the global path). It is a nice add on, but shouldn't be default behavior